File: 2024/6/alt-parallel.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Position [2]int
type Direction string

var positionsMapping = map[Direction][2]int{
	"^": {-1, 0},
	"<": {0, -1},
	">": {0, 1},
	"v": {1, 0},
}

var moveMapping = map[Direction]Direction{
	"^": ">",
	"<": "^",
	">": "v",
	"v": "<",
}

type GameState struct {
	obstacles      map[string]bool
	areaLimits     map[string]bool
	guardPosition  Position
	guardDirection Direction
}

func initialise(grid [][]string) GameState {
	obstacles := make(map[string]bool)
	areaLimits := make(map[string]bool)
	var guardPosition Position
	var guardDirection Direction

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			cell := grid[i][j]
			key := fmt.Sprintf("%d_%d", i, j)

			if cell == "#" {
				obstacles[key] = true
			} else if i == len(grid)-1 || j == len(grid[0])-1 || i == 0 || j == 0 {
				areaLimits[key] = true
			}

			if cell == "^" || cell == "<" || cell == ">" || cell == "v" {
				guardPosition = Position{i, j}
				guardDirection = Direction(cell)
			}
		}
	}

	return GameState{
		obstacles:      obstacles,
		areaLimits:     areaLimits,
		guardPosition:  guardPosition,
		guardDirection: guardDirection,
	}
}

func simulate(start Position, direction Direction, newObstacle string, areaLimits, obstacles map[string]bool) bool {
	localMoves := make(map[string]map[Direction]bool)
	pos := start
	dir := direction

	for {
		key := fmt.Sprintf("%d_%d", pos[0], pos[1])
		if localMoves[key] != nil {
			if localMoves[key][dir] {
				return true
			}
		} else {
			localMoves[key] = make(map[Direction]bool)
		}
		localMoves[key][dir] = true

		nextPos := Position{
			pos[0] + positionsMapping[dir][0],
			pos[1] + positionsMapping[dir][1],
		}
		nextKey := fmt.Sprintf("%d_%d", nextPos[0], nextPos[1])

		if areaLimits[nextKey] && nextKey != newObstacle {
			return false
		}

		if obstacles[nextKey] || nextKey == newObstacle {
			dir = moveMapping[dir]
		} else {
			pos = nextPos
		}
	}
}

func part1(grid [][]string) int {
	state := initialise(grid)
	visited := make(map[string]bool)

	for {
		key := fmt.Sprintf("%d_%d", state.guardPosition[0], state.guardPosition[1])
		if state.areaLimits[key] {
			visited[key] = true
			break
		}

		visited[key] = true
		nextPosition := Position{
			state.guardPosition[0] + positionsMapping[state.guardDirection][0],
			state.guardPosition[1] + positionsMapping[state.guardDirection][1],
		}

		nextKey := fmt.Sprintf("%d_%d", nextPosition[0], nextPosition[1])
		if state.obstacles[nextKey] {
			state.guardDirection = moveMapping[state.guardDirection]
		}

		state.guardPosition = Position{
			state.guardPosition[0] + positionsMapping[state.guardDirection][0],
			state.guardPosition[1] + positionsMapping[state.guardDirection][1],
		}
	}

	return len(visited)
}

func part2(grid [][]string) int {
	state := initialise(grid)
	startingPosition := state.guardPosition
	startingDirection := state.guardDirection
	visited := make(map[string]bool)

	for {
		key := fmt.Sprintf("%d_%d", state.guardPosition[0], state.guardPosition[1])
		if state.areaLimits[key] {
			visited[key] = true
			break
		}

		visited[key] = true
		nextPosition := Position{
			state.guardPosition[0] + positionsMapping[state.guardDirection][0],
			state.guardPosition[1] + positionsMapping[state.guardDirection][1],
		}

		nextKey := fmt.Sprintf("%d_%d", nextPosition[0], nextPosition[1])
		if state.obstacles[nextKey] {
			state.guardDirection = moveMapping[state.guardDirection]
		}

		state.guardPosition = Position{
			state.guardPosition[0] + positionsMapping[state.guardDirection][0],
			state.guardPosition[1] + positionsMapping[state.guardDirection][1],
		}
	}

	type simulationResult struct {
		cell   string
		isLoop bool
	}
	resultChan := make(chan simulationResult)
	activeSimulations := 0

	for cell := range visited {
		if cell == fmt.Sprintf("%d_%d", startingPosition[0], startingPosition[1]) {
			continue
		}

		activeSimulations++
		go func(cell string) {
			result := simulate(startingPosition, startingDirection, cell, state.areaLimits, state.obstacles)
			resultChan <- simulationResult{cell: cell, isLoop: result}
		}(cell)
	}

	newObstructions := make(map[string]bool)
	for i := 0; i < activeSimulations; i++ {
		result := <-resultChan
		if result.isLoop {
			newObstructions[result.cell] = true
		}
	}

	return len(newObstructions)
}

func readData(filename string) [][]string {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var grid [][]string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, strings.Split(line, ""))
	}

	return grid
}

func main() {
	grid := readData("data.txt")
	fmt.Printf("Part 1: %d\n", part1(grid))
	fmt.Printf("Part 2: %d\n", part2(grid))
}

File: 2024/6/alt-parallel.rs
use std::fs;
use std::path::Path;
use std::error::Error;
use std::collections::{HashMap, HashSet};
use tokio;
use rayon::prelude::*;

type Grid = Vec<Vec<char>>;
type Position = (i32, i32);

#[derive(Clone)]
struct State {
    obstacles: HashSet<String>,
    area_limits: HashSet<String>,
    guard_position: Position,
    guard_direction: char,
}

async fn read_data() -> Result<Vec<String>, Box<dyn Error>> {
    let contents = fs::read_to_string(Path::new("data.txt"))?;
    Ok(contents.lines().map(String::from).collect())
}

async fn format_data(rows: Vec<String>) -> Grid {
    rows.iter()
        .map(|row| row.chars().collect())
        .collect()
}

fn get_position_offset(direction: char) -> Position {
    match direction {
        '^' => (-1, 0),
        '<' => (0, -1),
        '>' => (0, 1),
        'v' => (1, 0),
        _ => panic!("Invalid direction"),
    }
}

fn get_next_direction(direction: char) -> char {
    match direction {
        '^' => '>',
        '<' => '^',
        '>' => 'v',
        'v' => '<',
        _ => panic!("Invalid direction"),
    }
}

fn init(grid: &Grid) -> State {
    let mut obstacles = HashSet::new();
    let mut area_limits = HashSet::new();
    let mut guard_position = (0, 0);
    let mut guard_direction = '^';

    for (i, row) in grid.iter().enumerate() {
        for (j, &cell) in row.iter().enumerate() {
            let pos = format!("{}_{}", i, j);
            
            if cell == '#' {
                obstacles.insert(pos);
            } else if i == 0 || j == 0 || i == grid.len() - 1 || j == row.len() - 1 {
                area_limits.insert(pos);
            }

            if "^<>v".contains(cell) {
                guard_position = (i as i32, j as i32);
                guard_direction = cell;
            }
        }
    }

    State {
        obstacles,
        area_limits,
        guard_position,
        guard_direction,
    }
}

fn simulate(
    start: Position,
    direction: char,
    new_obstacle: &str,
    area_limits: &HashSet<String>,
    obstacles: &HashSet<String>,
) -> bool {
    let mut local_moves: HashMap<String, HashSet<char>> = HashMap::new();
    let mut pos = start;
    let mut dir = direction;

    loop {
        let key = format!("{}_{}", pos.0, pos.1);
        if let Some(moves) = local_moves.get(&key) {
            if moves.contains(&dir) {
                return true;
            }
        }

        local_moves
            .entry(key)
            .or_insert_with(HashSet::new)
            .insert(dir);

        let offset = get_position_offset(dir);
        let next_pos = (pos.0 + offset.0, pos.1 + offset.1);
        let next_key = format!("{}_{}", next_pos.0, next_pos.1);

        if area_limits.contains(&next_key) && next_key != new_obstacle {
            return false;
        }

        if obstacles.contains(&next_key) || next_key == new_obstacle {
            dir = get_next_direction(dir);
        } else {
            pos = next_pos;
        }
    }
}

fn get_initial_path(state: &State) -> HashSet<String> {
    let mut visited = HashSet::new();
    let mut current_state = state.clone();

    while !state.area_limits.contains(&format!("{}_{}", current_state.guard_position.0, current_state.guard_position.1)) {
        visited.insert(format!("{}_{}", current_state.guard_position.0, current_state.guard_position.1));

        let offset = get_position_offset(current_state.guard_direction);
        let next_position = (
            current_state.guard_position.0 + offset.0,
            current_state.guard_position.1 + offset.1,
        );

        if current_state.obstacles.contains(&format!("{}_{}", next_position.0, next_position.1)) {
            current_state.guard_direction = get_next_direction(current_state.guard_direction);
        }

        let offset = get_position_offset(current_state.guard_direction);
        current_state.guard_position = (
            current_state.guard_position.0 + offset.0,
            current_state.guard_position.1 + offset.1,
        );
    }

    visited.insert(format!("{}_{}", current_state.guard_position.0, current_state.guard_position.1));
    visited
}

fn part1(grid: Grid) -> usize {
    let state = init(&grid);
    let visited = get_initial_path(&state);
    println!("Part 1: {}", visited.len());
    visited.len()
}

fn part2(grid: Grid) -> usize {
    let state = init(&grid);
    let visited = get_initial_path(&state);
    
    // Convert visited to Vec for parallel iteration
    let visited_vec: Vec<String> = visited
        .into_iter()
        .filter(|cell| {
            let coords: Vec<i32> = cell
                .split('_')
                .map(|x| x.parse().unwrap())
                .collect();
            (coords[0], coords[1]) != state.guard_position
        })
        .collect();

    // Parallel simulation
    let new_obstructions: HashSet<String> = visited_vec
        .par_iter()
        .filter(|&cell| {
            simulate(
                state.guard_position,
                state.guard_direction,
                cell,
                &state.area_limits,
                &state.obstacles,
            )
        })
        .cloned()
        .collect();

    println!("Part 2: {}", new_obstructions.len());
    new_obstructions.len()
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let data = read_data().await?;
    let formatted = format_data(data).await;
    part1(formatted.clone());
    part2(formatted);
    Ok(())
}